### PR TITLE
feat(pipelines): add evalhub pipelines and small tweaks to existing lora pipelines

### PR DESCRIPTION
**Pull request by MStokluska**
**Description of your changes:**

Added:

- lora minimal and full with evalhub optional (with kserve)
- sft minimal and full with evalhub optional (with kserve)
- osft minimal and full with evalhub optional (with kserve)

Edits:
- lora pipeline was allowing for MLflow params - this is not yet supported  

**Checklist:**

### Pre-Submission Checklist

- [ ] All tests and CI checks pass
- [ ] Pre-commit hooks pass without errors
- [ ] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [ ] The title for your pull request (PR) should follow our title convention.
  [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### Additional Checklist Items for New or Updated Components/Pipelines

- [ ] `metadata.yaml` includes fresh `lastVerified` timestamp
- [ ] All [required files](https://github.com/kubeflow/pipelines-components/blob/main/docs/CONTRIBUTING.md#required-files)
  are present and complete
- [ ] OWNERS file lists appropriate maintainers
- [ ] README provides clear documentation with usage examples
- [ ] Component follows `snake_case` naming convention
- [ ] No security vulnerabilities in dependencies
- [ ] Containerfile included if using a custom base image

## Summary by CodeRabbit

* **New Features**
  * Eval Hub evaluation via KServe added.
  * New finetuning+evaluation pipelines: LoRA, OSFT, SFT and “minimal” variants.

* **Bug Fixes / Changes**
  * MLflow-related configuration parameters removed from LoRA training interfaces.

* **Documentation**
  * Added docs for Eval Hub evaluation, KServe evaluator, and new pipelines.

* **Tests**
  * Added unit tests for the Eval Hub KServe evaluator and pipeline compilation.

* **Chores**
  * Package metadata updated to include new evaluation and pipeline modules.

### CODE REVIEW

File: components/evaluation/evalhub_kserve/component.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+            if pvc.get("claimName"):
+                pvc_volumes[v["name"]] = pvc["claimName"]
+
+        containers = pod_spec.get("spec", {}).get("containers", [])
+        for c in containers:
+            for vm in c.get("volumeMounts", []):
+                vol_name = vm["name"]
+                mount_path = vm["mountPath"]
+                if vol_name in pvc_volumes and model_path.startswith(mount_path):
+                    return pvc_volumes[vol_name], mount_path
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply components/evaluation/evalhub_kserve/component.py]: will fix in next commit _find_workspace_pvc will now normalize both paths with a trailing / before comparison, so /mnt/models2/... no longer falsely matches /mnt/models

File: components/evaluation/evalhub_kserve/component.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+                "containers": [{
+                    "name": "kserve-container",
+                    "image": image,
+                    "command": ["python", "-m", "vllm.entrypoints.openai.api_server"],
+                    "args": [
+                        "--port=8080",
+                        "--model=/mnt/models",
+                        f"--served-model-name={served_model_name}",
+                        "--trust-remote-code",
+                    ],
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply components/evaluation/evalhub_kserve/component.py]: Going to add trust_remote_code: bool = False parameter to evalhub_evaluator_kserve, threaded it through _create_serving_runtime, and --trust-remote-code is now only appended when explicitly enabled.

File: components/evaluation/evalhub/kserve/component.py [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+                return url
+        fallback = f"http://{name}-predictor.{namespace}.svc.cluster.local:8080/v1"
+        logger.warning(f"Could not read ISVC status, falling back to: {fallback}")
+        return fallback
+
+    def _cleanup_kserve(namespace, sr_name, isvc_name):
+        logger.info(f"Cleaning up InferenceService {isvc_name} and ServingRuntime {sr_name}")
+        _k8s_api("DELETE", f"{KSERVE_ISVC_API}/namespaces/{namespace}/inferenceservices/{isvc_name}")
+        _k8s_api("DELETE", f"{KSERVE_SR_API}/namespaces/{namespace}/servingruntimes/{sr_name}")
+        logger.info(f"Cleanup complete for {isvc_name} / {sr_name}")
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply components/evaluation/evalhub/kserve/component.py:330]: Will fix - _cleanup_kserve now checks each delete response and logs a warning on non-404 failures. "Cleanup complete" still logs at the end for audit, but failures are no longer silent.

File: components/evaluation/evalhub/kserve/component.py [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+        elif os.path.isdir(model_artifact.path):
+            final_model_path = model_artifact.path
+            logger.info(f"Using model from artifact path: {final_model_path}")
+
+    if final_model_path is None and model_path:
+        final_model_path = model_path
+        logger.info(f"Using model_path: {final_model_path}")
+
+    if final_model_path is None:
+        raise ValueError("No model provided. Supply model_artifact or model_path.")
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply components/evaluation/evalhub/kserve/component.py:365]: Updating the docstring to clarify model_path accepts only local filesystem paths, not HuggingFace IDs. Remote model serving is out of scope for this KServe-PVC component.

File: components/evaluation/evalhub_kserve/component.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+            eval_config["benchmarks"] = parsed
+        else:
+            raise ValueError("Provide either 'benchmarks' list or 'collection_id'.")
+
+        submit_url = f"{evalhub_url.rstrip('/')}/api/v1/evaluations/jobs"
+        logger.info(f"Submitting evaluation job to {submit_url}")
+        logger.info(f"Config: {json.dumps(eval_config, indent=2)}")
+
+        resp = requests.post(submit_url, json=eval_config, headers=headers, timeout=30, verify=False)
+        if resp.status_code not in (200, 201, 202):
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply components/evaluation/evalhub_kserve/component.py]: adding verify_tls: bool = False parameter. All three Eval Hub HTTP calls (POST, GET, DELETE) now use it instead of hardcoded verify=False. Defaults to False for in-cluster self-signed certs, but can be enabled

File: components/evaluation/evalhub_kserve/README.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+     --serviceaccount=<namespace>:<pipeline-sa> \
+     -n <namespace>
+   ```
+
+4. **Workspace PVC** must use `ReadWriteMany` access mode (NFS-backed) so the KServe
+   predictor pod can mount the fine-tuned model.
+
+5. **`kubernetes-credentials` Secret** containing `KUBERNETES_SERVER_URL` and
+   `KUBERNETES_AUTH_TOKEN` for K8s API access from within the component.
+
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply components/evaluation/evalhub_kserve/README.md]: updating all 6 evalhub pipeline docstrings to say "The eval component uses the in-cluster ServiceAccount token for K8s API access" instead of referencing the kubernetes-credentials Secret. That secret is only used by the training component, not the eval component.

File: pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+        ),
+    ),
+)
+def lora_pipeline_evalhub_easy(
+    # =========================================================================
+    # KEY PARAMETERS (Required/Important) - Sorted by step
+    # =========================================================================
+    phase_01_dataset_man_data_uri: str,
+    phase_03_eval_opt_evalhub_url: str = "",
+    phase_02_train_man_train_batch: int = 128,
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py:52]: The dataset_download component already validates URI schemes (accepts only hf://, s3://, https://) and the evalhub_url is an internal API endpoint

File: pipelines/training/finetuning_evalhub/sft/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+from components.deployment.kubeflow_model_registry import kubeflow_model_registry
+from components.evaluation.evalhub_kserve import evalhub_evaluator_kserve
+from components.training.finetuning.sft import train_model
+
+# =============================================================================
+# PVC Configuration (COMPILE-TIME settings)
+# =============================================================================
+PVC_SIZE = "50Gi"
+PVC_STORAGE_CLASS = "nfs-csi"
+PVC_ACCESS_MODES = ["ReadWriteMany"]
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply pipelines/training/finetuning_evalhub/sft/pipeline.py:24]: PVC_STORAGE_CLASS is a compile-time constant across all pipelines in this repo (LoRA, SFT, OSFT). KFP workspace PVC config is set at compile time via dsl.PipelineConfig and cannot be parameterized at runtime.

File: pipelines/training/finetuning_evalhub/sft/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+                }
+            ),
+        ),
+    ),
+)
+def sft_pipeline_evalhub(
+    # =========================================================================
+    # KEY PARAMETERS (Required/Important) - Sorted by step
+    # =========================================================================
+    phase_01_dataset_man_data_uri: str,
```
[the remark this answers is not in the document]
Comment (MStokluska) [reply pipelines/training/finetuning_evalhub/sft/pipeline.py:48]: The dataset_download component is not part of this PR - it's a pre-existing shared component used by other pipelines in the repo.

File: pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        evalhub_poll_interval=30,
+        mlflow_experiment_name=phase_03_eval_opt_mlflow_experiment,
+        gpu_count=phase_03_eval_opt_kserve_gpu_count,
+        memory=phase_03_eval_opt_kserve_memory,
+        cpu=phase_03_eval_opt_kserve_cpu,
+    )
+    eval_task.set_caching_options(False)
+    kfp.kubernetes.set_image_pull_policy(eval_task, "IfNotPresent")
+
+    for _task in [dataset_download_task, training_task]:
```
Comment (Fiona-Waters) [inline pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py:227]: Would there be any use case where we might need to include eval_task for the hf token?

Comment (Fiona-Waters) [inline pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py:227]: Would there be any use case to include the eval task here too?

**Review by Fiona-Waters — COMMENTED** [review]
Successfully ran the lora_minimal_evalhub on the team cluster. Looks good to me. Just one question.

File: pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py (hunk 10 above)
Comment (MStokluska) [reply pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py:227]: Yes, good catch - some benchmarks do require HF token (however, this requirement is not in any of the default ones.
So for minimal flavor, I'm going to leave that out since the benchmarks aren't configurable - for extended, I'll add the eval to tasks. Thanks!

Comment (MStokluska) [reply pipelines/training/finetuning_evalhub/lora_minimal/pipeline.py:227]: As above, not for minimal but for full flavor we need that, thanks!

**Review by Fiona-Waters — APPROVED** [review]
/lgtm
/approve

**Comment by MStokluska** [comment]
@nsingla @HumairAK could we get approval on this (or review if required?)

**Review by Fiona-Waters — APPROVED** [review]

**Comment by Fiona-Waters** [comment]
@nsingla I'm trying to merge this but it looks like I have read only permissions. Could you please merge it and we can check the permissions issues later?

**Review by nsingla — APPROVED** [review]
/lgtm
